modules/s3/src/main/clojure/xtdb/s3.clj

```diff
@@ -48,6 +46,72 @@
                             (catch NoSuchKeyException _
                               (throw (os/obj-missing-exception k))))))))
 
+(def multipart-size-threshold (* 100 1024 1024))
+(def chunk-size (* 10 1024 1024))
+
+(defn single-object-upload
+  [{:keys [^S3AsyncClient client ^S3Configurator configurator bucket prefix]} k ^ByteBuffer buf]
+  (.putObject client
+              (-> (PutObjectRequest/builder)
+                  (.bucket bucket)
+                  (.key (str prefix k))
+                  (->> (.configurePut configurator))
+                  ^PutObjectRequest (.build))
+              (AsyncRequestBody/fromByteBuffer buf)))
+
+(defn multipart-upload
+  [{:keys [^S3AsyncClient client bucket prefix]} k ^ByteBuffer buf]
+  (let [prefixed-key (str prefix k)
+        initiate-request (-> (CreateMultipartUploadRequest/builder)
+                             (.bucket bucket)
+                             (.key prefixed-key)
+                             ^CreateMultipartUploadRequest (.build))]
+    (-> (.createMultipartUpload client initiate-request)
+        (util/then-compose (fn [^CreateMultipartUploadResponse initiate-response]
```

Comment:
Thinking we could potentially include some try logic in this then-compose block which, on an exception, calls to abortMultipartUpload?

Reply:
It does seem like multipart operations do not automatically get cleaned up - thats a bucket level config: https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html

We could probably add that into our S3 stack - but might be better to allow XTDB to cleanup failed multipart uploads itself.